### PR TITLE
[fix] Revert change that removed the option to run OffloadModel with out activation checkpointing.

### PR DESCRIPTION
**Pull request by anj-s**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
- `checkpoint_activation` was removed incorrectly. Reverted SyncShard change since it makes the code more readable inspite of having another path. Working on code refactoring but wanted to get this change checked in the meantime.
- Added tests since we did not detect the missing codepath
- Modified benchmarks to account for the checkpoint_activation flag. Will be adding a benchmark in an upcoming PR once I have a few more run tabulated.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/experimental/nn/offload.py [hunk 1]
```diff
@@ -292,6 +292,75 @@ def backward(ctx, *grad_outputs):  # type: ignore
         return (None, None) + grads
 
 
+class ShardSyncLayer(torch.autograd.Function):
```
Comment (blefaudeux) [inline fairscale/experimental/nn/offload.py:295]: at least this part I'm a bit familiar with :)

File: fairscale/experimental/nn/offload.py [hunk 2]
```diff
@@ -386,4 +455,23 @@ def forward(self, *inputs: Any, **_: Any) -> Any:
 
         # We need the second param to be a dummy input to enable the
         # backward pass to be triggered for integer inputs.
-        return ActivationCheckpointing.apply(*inputs, torch.tensor([], requires_grad=True), self)
+        if self._checkpoint_activation:
```
Comment (blefaudeux) [inline fairscale/experimental/nn/offload.py:458]: oh, I must have reviewed the offending PR and missed that, sorry about that

File: tests/experimental/nn/test_offload.py [hunk 3]
```diff
@@ -32,20 +32,37 @@ def test_single_run():
...
+            checkpoint_activation=checkpoint_activation,
+        )
+        offload_optimizer = torch.optim.SGD(offload_model.parameters(), lr=0.001)
+
+        input = torch.ones(1000, 2).to(device)
+        labels = torch.ones(1000, 2).to(device)
+        offload_model.train()
+        pred = offload_model(input)
+        loss_fn = torch.nn.MSELoss(reduction="sum")
+        loss = loss_fn(pred, labels)
```
Comment (blefaudeux) [inline tests/experimental/nn/test_offload.py:51]: checking elsewhere for some form of parity ? wondering just in case

**Review by blefaudeux — APPROVED** [review]
LGTM especially since revert, else some missing bits of the big picture, checking 1:1

File: fairscale/experimental/nn/offload.py (hunk 2 above)
Comment (anj-s) [reply fairscale/experimental/nn/offload.py:458]: No worries! I realized that the tests weren't really catching this so glad I realized it.

**Pull request merged**